src/core/Object3D.js

```diff
@@ -575,6 +575,53 @@ class Object3D extends EventDispatcher {
 
 	}
 
+	*traverseGenerator( condition ) {
```

Comment:
It would be nice if we could add these methods without all the code duplication.

BTW: Enabling the usage of different coding pattern is fine but how is the performance of generators in the meanwhile? I remember it took a while until they got optimized but I wonder if the new methods are on par with the existing ones.

Reply:
I'm not sure how we could avoid the duplication here to be quite honest. Most of the business logic is inherently different due to the usage of generators so even though it might look similar (and be so for 2 lines), I fear we might incur overhead in trying to merge the regular and generator versions. I'm not the best JS developer though - far from it - so if you see a way to do so with no overhead I'd be interested.

Generators absolutely have worse performance in certain cases (last I remember for a full walk through a simple list of integers I was getting around +50% time with generators) but can be an invaluable time save if the calculation required per item is rather heavy. Not only that, but with generators we go from O(n) spacial complexity to O(1), which can be an interesting benefit in niche cases. Depending on the situation, going from the regular traversal to a generator-based one requires very little refactoring on the user's side for a possibly significant performance boost.
I'll run some performance tests again just in case and will comment the results.

Reply:
I've run two simple tests just as a proof of the easily obtainable performance benefits of generators in certain use cases.
Note that the `array` examples are not what the `traverse` method is doing since it does not return anything, but since this PR is about the benefits of stopping computation early in an ergonomic way I felt it was a more appropriate comparison (since afaik it's currently impossible to stop the traversal midway given that the callback's return value isn't used).
 
In a [simple case](https://jsbench.me/gmlus7oq9q) with barely any computation for each element the array version fairly obviously far outclasses the generator by about 60-80%.
However, in a [slightly more complex case](https://jsbench.me/g7lus9gsth) (that is still a far cry from real world applications) the time lost by the generator pattern is so negligible in comparison to the value mapping and checking that it ends up being ~15% faster.

Generators are not the right choice for all problems, but when they are the benefits are non-negligible.

Reply:
My feeling is that these methods are best added to THREE.SceneUtils:

```
import { traverseGenerator } from 'three/addons/utils/SceneUtils.js';

for (const child of traverseGenerator(object)) {

  // 

}
```

Then we can implement them without worrying about the code duplication or bundle size — it's not bundled if not used. And if this usage style becomes more popular, or JavaScript engines optimize iterator performance more in future versions, we can consider adding the methods to THREE.Object3D again.